homeassistant/components/upcloud.py

```diff
@@ -0,0 +1,133 @@
+"""
+Support for UpCloud.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/upcloud/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.util import Throttle
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['upcloud-api==0.3.9']
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_CORE_NUMBER = 'core_number'
+ATTR_HOSTNAME = 'hostname'
+ATTR_MEMORY_AMOUNT = 'memory_amount'
+ATTR_STATE = 'state'
+ATTR_TITLE = 'title'
+ATTR_UUID = 'uuid'
+ATTR_ZONE = 'zone'
+
+CONF_SERVERS = 'servers'
+
+DATA_UPCLOUD = 'data_upcloud'
+DOMAIN = 'upcloud'
+
+DEFAULT_COMPONENT_NAME = 'UpCloud {}'
+DEFAULT_COMPONENT_DEVICE_CLASS = 'power'
+
+UPCLOUD_PLATFORMS = ['binary_sensor', 'switch']
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the UpCloud component."""
+    import upcloud_api
+
+    conf = config[DOMAIN]
+    username = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
+
+    manager = upcloud_api.CloudManager(username, password)
+
+    try:
+        manager.authenticate()
+    except upcloud_api.UpCloudAPIError:
+        _LOGGER.error("Authentication failed.")
+        return False
+
+    hass.data[DATA_UPCLOUD] = UpCloud(manager)
+    hass.data[DATA_UPCLOUD].update()
```

Comment:
Calling update() here will delay HA launch
Can you find away around it ?

Reply:
Not sure. In cases where there are only switches _or_ binary sensors configured, this update() could just be removed, the component will call it whenever it needs to refresh. But when there are both switches _and_ binary sensors configured, one of them will end up being configured without the data (because of the throttling) on start and thus shows incorrect/missing data.

Reply:
Looking closer, the UpCloud object is shared by both switches and binary sensors, so the servers list (which is the only thing in update()) will be shared.